src/sync/app.cpp

```diff
@@ -0,0 +1,163 @@
+    return status >= 300;
+}
+
+void App::login_with_credentials(const std::shared_ptr<AppCredentials> credentials,
+                                 std::function<void(std::shared_ptr<SyncUser>, std::unique_ptr<error::AppError>)> completion_block) {
+    // construct the route
+    std::string route = util::format("%1/providers/%2/login", m_auth_route, provider_type_from_enum(credentials->m_provider));
+
+    auto handler = [&](const Response response) {
+        // if there is a already an error code, pass the error upstream
+        if (response_code_is_fatal(response.http_status_code)) { // FIXME: handle
+            return completion_block(nullptr, handle_error(response));
+        }
+
+        nlohmann::json json;
+        try {
+            json = nlohmann::json::parse(response.body);
+        } catch(...) {
+            return completion_block(nullptr, std::make_unique<error::AppError>(app::error::ClientError(app::error::ClientError::Code::bad_response)));
+        }
+
+        realm::SyncUserIdentifier identifier {
+            json["user_id"].get<std::string>(),
+            m_auth_route
+        };
+
+        std::shared_ptr<realm::SyncUser> sync_user;
+        try {
+            sync_user = realm::SyncManager::shared().get_user(identifier,
+                                                              json["refresh_token"].get<std::string>(),
+                                                              json["access_token"].get<std::string>());
+        } catch (const error::ClientError& err) {
+            return completion_block(nullptr, std::make_unique<error::ClientError>(err));
+        } catch (const error::ServiceError& err) {
+            return completion_block(nullptr, std::make_unique<error::ServiceError>(err));
+        } catch (const error::AppError& err) {
+            return completion_block(nullptr, std::make_unique<error::AppError>(err));
+        }
+
+        std::string profile_route = util::format("%1/auth/profile", m_base_route);
+        std::string bearer = util::format("Bearer %1", sync_user->access_token());
+
+        GenericNetworkTransport::get()->send_request_to_server({
+            Method::get,
+            profile_route,
+            m_request_timeout_ms,
+            {
+                { "Content-Type", "application/json;charset=utf-8" },
+                { "Accept", "application/json" },
+                { "Authorization", bearer}
+            },
+            std::string()
+        }, [&](const Response profile_response) {
+            // if there is a already an error code, pass the error upstream
+            if (response_code_is_fatal(profile_response.http_status_code)) {
+                return completion_block(nullptr, handle_error(profile_response));
+            }
+
+            json = nlohmann::json::parse(profile_response.body);
+
+            std::vector<SyncUserIdentity> identities;
+            auto identities_json = json["identities"];
```

Comment:
Relying on the JSON being well-formed seems a bit dangerous? I would wrap this and the below code in some sort of try-catch?